plugin-bsp/src/main/kotlin/org/jetbrains/plugins/bsp/impl/server/connection/DefaultBspConnection.kt

```diff
@@ -120,124 +128,117 @@ private class CancelableInvocationHandlerWithTimeout(
-    )
+      bspSyncConsole.startTask(
+        taskId = CONNECT_TASK_ID,
+        title = BspPluginBundle.message("console.task.auto.connect.title"),
+        message = BspPluginBundle.message("console.task.auto.connect.in.progress"),
+        cancelAction = { coroutineContext.cancel() },
+      )
 
-    try {
-      connectOrThrowIfFailed(bspSyncConsole, taskId)
-    } catch (e: Exception) {
-      if (e is CancellationException) {
-        bspSyncConsole.finishSubtask(
-          subtaskId = CONNECT_SUBTASK_ID,
-          message = BspPluginBundle.message("console.subtask.connect.cancelled"),
-          result = FailureResultImpl(),
-        )
-      } else {
-        bspSyncConsole.finishSubtask(
-          subtaskId = CONNECT_SUBTASK_ID,
-          message = BspPluginBundle.message("console.subtask.connect.failed"),
-          result = FailureResultImpl(e),
-        )
-        throw e
+      try {
+        connectOrThrowIfFailed()
+        bspSyncConsole.finishTask(CONNECT_TASK_ID, BspPluginBundle.message("console.task.auto.connect.success"))
+      } catch (e: Exception) {
+        if (e is CancellationException) {
+          bspSyncConsole.finishTask(
+            taskId = CONNECT_TASK_ID,
+            message = BspPluginBundle.message("console.task.auto.connect.cancelled"),
+            result = SkippedResultImpl(),
+          )
+        } else {
+          bspSyncConsole.finishTask(
+            taskId = CONNECT_TASK_ID,
+            message = BspPluginBundle.message("console.task.auto.connect.failed"),
+            result = FailureResultImpl(e),
+          )
+          throw e
+        }
       }
     }
   }
 
-  private suspend fun connectOrThrowIfFailed(bspSyncConsole: TaskConsole, taskId: Any) {
+  private suspend fun connectOrThrowIfFailed() {
     val bspClient = createBspClient()
-    bspSyncConsole.addMessage(
-      taskId = CONNECT_SUBTASK_ID,
-      message = BspPluginBundle.message("console.task.connect.message.in.progress"),
-    )
     val inMemoryConnection =
       BspServerProvider.getBspServer()?.getConnection(
         project,
         null,
         bspClient,
       )
     if (inMemoryConnection != null) {
-      val console = BspConsoleService.getInstance(project).bspSyncConsole
-      connectBuiltIn(console, inMemoryConnection)
+      connectBuiltIn(inMemoryConnection)
     } else {
-      val newConnectionDetails = connectionDetailsProviderExtension.provideNewConnectionDetails(project, null)
-
+      val newConnectionDetails = connectionDetailsProviderExtension.provideNewConnectionDetails(project, connectionDetails)
       if (newConnectionDetails != null) {
-        this.connectionDetails = newConnectionDetails
-        newConnectionDetails.connect(bspSyncConsole, taskId, bspClient)
-      } else {
-        error("Cannot connect. Please check your connection file.")
+        connectionDetails = newConnectionDetails
+        newConnectionDetails.connect(bspClient)
+      } else if (!isConnected()) {
```

Comment:
**suggestion:** Consider adding specific error handling for connection failures

The old version had a specific error message for when connection details couldn't be provided. Adding this back could be helpful for debugging and error reporting.